daml-lf/validation/src/main/scala/com/digitalasset/daml/lf/validation/Typing.scala

```diff
@@ -641,35 +641,92 @@ private[validation] object Typing {
       case otherwise => throw EPatternTypeMismatch(ctx, otherwise, scrutType)
     }
 
-    private[this] def introPatternForType(scrutType: Type): CasePat => Env =
-      scrutType match {
+    private[this] def addPatternRank(ranks: Set[Int], pat: CasePat): MatchedRanks =
+      pat match {
+        case CPVariant(tycon, variant, _) =>
+          lookupDataType(ctx, tycon) match {
+            case DDataType(_, _, data: DataVariant) =>
+              SomeRanks(ranks + data.constructorRank(variant))
+            case _ =>
+              throw EUnknownDefinition(ctx, LEDataVariant(tycon))
+          }
+        case CPEnum(tycon, constructor) =>
+          lookupDataType(ctx, tycon) match {
+            case DDataType(_, _, data: DataEnum) =>
+              SomeRanks(ranks + data.constructorRank(constructor))
+            case _ =>
+              throw EUnknownDefinition(ctx, LEDataEnum(tycon))
+          }
+        case CPPrimCon(pc) =>
+          pc match {
+            case PCFalse | PCUnit => SomeRanks(ranks + 1)
+            case PCTrue => SomeRanks(ranks + 0)
```

Comment:
Based on the definition of the expected patterns below, I think it should be this:

```suggestion
            case PCFalse | PCUnit => SomeRanks(ranks + 0)
            case PCTrue => SomeRanks(ranks + 1)
```